tpm/src/emulator.rs

```diff
@@ -0,0 +1,470 @@
+// Copyright © 2022, Microsoft Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use crate::ioctl::{
+    Commands, MemberType, Ptm, PtmCap, PtmEst, PtmInit, PtmResult, PtmSetBufferSize,
+};
+use crate::ioctl::{TPM_CRB_BUFFER_MAX, TPM_SUCCESS};
+use crate::socket::SocketDev;
+use anyhow::anyhow;
+use libc::c_void;
+use libc::{sockaddr_storage, socklen_t};
+use std::convert::TryInto;
+use std::os::unix::io::RawFd;
+use std::path::Path;
+use std::{mem, ptr};
+use thiserror::Error;
+
+const TPM_REQ_HDR_SIZE: usize = 10;
+
+/* capability flags returned by PTM_GET_CAPABILITY */
+const PTM_CAP_INIT: u64 = 1;
+const PTM_CAP_SHUTDOWN: u64 = 1 << 1;
+const PTM_CAP_GET_TPMESTABLISHED: u64 = 1 << 2;
+const PTM_CAP_SET_LOCALITY: u64 = 1 << 3;
+const PTM_CAP_CANCEL_TPM_CMD: u64 = 1 << 5;
+const PTM_CAP_RESET_TPMESTABLISHED: u64 = 1 << 7;
+const PTM_CAP_STOP: u64 = 1 << 10;
+const PTM_CAP_SET_DATAFD: u64 = 1 << 12;
+const PTM_CAP_SET_BUFFERSIZE: u64 = 1 << 13;
+
+///Check if the input command is selftest
+///
+pub fn tpm_util_is_selftest(input: Vec<u8>, in_len: usize) -> bool {
+    if in_len >= TPM_REQ_HDR_SIZE {
+        let ordinal: &[u8; 4] = input[6..6 + 4]
+            .try_into()
+            .expect("tpm_util_is_selftest: slice with incorrect length");
+
+        return u32::from_ne_bytes(*ordinal).to_be() == 0x143;
+    }
+    false
+}
+
+#[derive(Error, Debug)]
+pub enum TPMEmuError {
+    #[error("Input socket path for TPM Emulator does not exist")]
+    TPMSocketPathExists(),
+    #[error("Could not initialize TPM Emulator Backend")]
+    InitializeTPMEmulator(#[source] anyhow::Error),
+    #[error("Failed to create data fd to pass to swtpm")]
+    PrepareDataFd(#[source] anyhow::Error),
+    #[error("Failed to run Ctrl Cmd: {0}")]
+    RunCtrlCmd(#[source] anyhow::Error),
+    #[error("TPM Emulator doesn't implement min required capabilities: {0}")]
+    CheckCaps(#[source] anyhow::Error),
+    #[error("TPM Emulator failed to deliver request: {0}")]
+    DeliverRequest(#[source] anyhow::Error),
+    #[error("TPM Emulator failed to send/receive msg on data fd: {0}")]
+    SendReceive(#[source] anyhow::Error),
+}
+
+type Result<T> = anyhow::Result<T, TPMEmuError>;
+
+#[derive(Clone)]
+pub struct TPMBackendCmd {
+    pub locty: u8,
+    pub input: Vec<u8>,
+    pub input_len: usize,
+    pub output: Vec<u8>,
+    pub output_len: usize,
+    pub selftest_done: bool,
+}
+
+pub struct TPMEmulator {
```

Comment:
This commit is the same:

* Don't repeat the context of the types, we know this a TPM emulator as it is in the TPM crate hiearchy.
* Inconsistency with nomenclature (this makes it much harder to search the code)